google/cloud/spanner_dbapi/connection.py

```diff
@@ -175,43 +177,56 @@ def _rerun_previous_statements(self):
         from the last transaction.
         """
         for statement in self._statements:
-            res_iter, retried_checksum = self.run_statement(statement, retried=True)
-            # executing all the completed statements
-            if statement != self._statements[-1]:
-                for res in res_iter:
-                    retried_checksum.consume_result(res)
-
-                _compare_checksums(statement.checksum, retried_checksum)
-            # executing the failed statement
+            if isinstance(statement, list):
+                statements, checksum = statement
+
+                transaction = self.transaction_checkout()
```

Comment:
The `transaction_checkout()` method uses `_session_checkout()` method to get session:
https://github.com/googleapis/python-spanner/blob/916ec7ac91945b50cc3dcb12792f3912cce07482/google/cloud/spanner_dbapi/connection.py#L226

And `_session_checkout()` method checks if a session object is already obtained by the connection. If not, it'll be taken from the pool, otherwise the previously taken session will be used:
https://github.com/googleapis/python-spanner/blob/916ec7ac91945b50cc3dcb12792f3912cce07482/google/cloud/spanner_dbapi/connection.py#L136

So, it looks like it should work fine. I'll take a closer look at the tests, probably it already checked somewhere - if not, I think we can add a check.

Reply:
You're right, the `_session_checkout()` function seems to handle that correctly.